rust/kernel/thread.rs

```diff
@@ -0,0 +1,404 @@
+    ///
+    /// This function actually doesn't dereference `arg` or call `f`, so even if
+    /// the users pass incorrect parameters this function won't run into
+    /// trouble. But if the users provide incorrect `arg` or `f` the new
+    /// thread will corrupt memory or do other unsafe behaviors, so
+    /// make it `unsafe`. Otherwise, it is [`Thread::wake_up`] that should be
+    /// made as `unsafe` because it will trigger the call of the unsafe function
+    /// `f`, this is undesirable given that [`Thread::wake_up`] on a
+    /// [`Thread::try_new`] created thread is safe.
+    ///
+    /// The safety requirements of calling this function are:
+    ///
+    /// - Make sure `arg` is a proper pointer that points to a valid memory
+    ///   location when the new thread begins to run.
+    ///
+    /// - Make sure `f` is a proper function pointer and `f` handles `arg`
+    ///   correctly.
+    ///
+    /// # Context
+    ///
+    /// This function might sleep in `kthread_create_on_node` due to the memory
+    /// allocation and waiting for the completion, therefore do not call this
+    /// in atomic contexts (i.e. preemption-off contexts).
+    pub unsafe fn try_new_extern(
+        name: CStr,
+        f: unsafe extern "C" fn(*mut c_types::c_void) -> c_types::c_int,
+        arg: *mut c_types::c_void,
+    ) -> KernelResult<Self> {
+        let task;
+
+        // SAFETY: `kthread_create_on_node` will copy the content of `name`,
+        // so we don't need to make the `name` live longer.
+        task = ptr_to_result(bindings::kthread_create_on_node(
+            Some(f),
+            arg,
+            bindings::NUMA_NO_NODE,
+            cstr!("%s").as_ptr() as _,
+            name.as_ptr(),
+        ))?;
+
+        // Increases the refcount of the task, so that it won't go away if it
+        // `do_exit`.
+        //
+        // SAFETY: `task` is a proper pointer pointing to a newly created
+        // thread.
+        rust_helper_get_task_struct(task);
+
+        Ok(Thread { task })
+    }
+
+    /// Creates a new thread.
+    ///
+    /// To use this function, you need to provide a `ThreadFunctor<A>` type,
+    /// as the generic type argument for the funciton.
+    ///
+    ///
+    /// Note that we have constraint for `arg` as `'static`, this is because
+    /// the new thread (which actually uses the `arg`) may outlive the caller
+    /// of the function. The new thread may run until the end of the whole
+    /// system, hence the `'static` lifetime.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use alloc::boxed::Box;
+    /// use core::fmt::Debug;
+    /// use kernel::thread::{Thread, ThreadFunctor};
+    ///
+    /// struct SimpleBoxFunc;
+    ///
+    /// impl<T> ThreadFunctor<Box<T>> for SimpleBoxFunc
+    /// where
+    ///     T: Send + Debug,
+    /// {
+    ///     // `Box<T>` implements [`ThreadArg`]
+    ///     fn func(b: Box<T>) -> KernelResult {
+    ///         println!("I got {:?}", b);
```

Comment:
`info!` instead now.

I would avoid using `Debug` here too.

Reply:
Got it, I will change to `info!` and remove `Debug`.